gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java

```diff
@@ -1,154 +1,148 @@
+public class JobCatalogListenersList implements JobCatalogListener, JobCatalogListenersContainer {
+  private final CallbacksDispatcher<JobCatalogListener> _disp;
 
   public JobCatalogListenersList() {
     this(Optional.<Logger>absent());
   }
 
   public JobCatalogListenersList(Optional<Logger> log) {
-    _log = log.isPresent() ? log.get() : LoggerFactory.getLogger(getClass());
-    _debugLogEnabled = _log.isDebugEnabled();
+    _disp = new CallbacksDispatcher<JobCatalogListener>(Optional.<ExecutorService>absent(), log);
+  }
+
+  public Logger getLog() {
+    return _disp.getLog();
   }
 
   public synchronized List<JobCatalogListener> getListeners() {
-    return new ArrayList<>(_listeners);
+    return _disp.getListeners();
   }
 
+  @Override
   public synchronized void addListener(JobCatalogListener newListener) {
-    Preconditions.checkNotNull(newListener);
-    _listeners.add(newListener);
+    _disp.addListener(newListener);
   }
 
+  @Override
   public synchronized void removeListener(JobCatalogListener oldListener) {
-    Preconditions.checkNotNull(oldListener);
-    _listeners.remove(oldListener);
+    _disp.removeListener(oldListener);
   }
 
   @Override
   public synchronized void onAddJob(JobSpec addedJob) {
     Preconditions.checkNotNull(addedJob);
-    callbackAllListeners(new AddJobCallback(addedJob));
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new AddJobCallback(addedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onAddJob interrupted.");
+    }
   }
 
   @Override
   public synchronized void onDeleteJob(JobSpec deletedJob) {
     Preconditions.checkNotNull(deletedJob);
-    callbackAllListeners(new DeleteJobCallback(deletedJob));
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new DeleteJobCallback(deletedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onDeleteJob interrupted.");
+    }
   }
 
   @Override
   public synchronized void onUpdateJob(JobSpec originalJob, JobSpec updatedJob) {
     Preconditions.checkNotNull(originalJob);
     Preconditions.checkNotNull(updatedJob);
-    callbackAllListeners(new UpdateJobCallback(originalJob, updatedJob));
-  }
-
-  public void callbackAllListeners(Callback callback) {
-    for (JobCatalogListener listener: _listeners) {
-      callbackOneListener(callback, listener);
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new UpdateJobCallback(originalJob, updatedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onUpdateJob interrupted.");
     }
   }
 
-  public void callbackOneListener(Callback callback, JobCatalogListener listener) {
-    String callbackMsg = null;
-    if (_debugLogEnabled) {
-      callbackMsg = "callback " + callback + " on " + listener;
-      _log.debug("Started: " + callbackMsg);
-    }
+  public void callbackOneListener(Function<JobCatalogListener, Void> callback,
```

Comment:
This is needed for the contract, that when we add a new listener, it should get callbacks for all existing jobs. See InMemoryJobCatalog.addListener() .